### PR TITLE
Annotate implementations of IEventSymbol

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for review

**Comment by 333fred** [comment]
>         internal EventSymbol GetLeastOverriddenEvent(NamedTypeSymbol accessingTypeOpt)

Should be annotated.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs:167 in 4639bb4. [](commit_id = 4639bb49a2bb56b7f678c9c8ddc6ce235d55708f, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs [hunk 1]
```diff
@@ -258,14 +261,16 @@ internal EventSymbol AsMember(NamedTypeSymbol newOwner)
         {
             Debug.Assert(this.IsDefinition);
             Debug.Assert(ReferenceEquals(newOwner.OriginalDefinition, this.ContainingSymbol.OriginalDefinition));
+#nullable disable // Can 'newOwner as SubstitutedNamedTypeSymbol' be null? https://github.com/dotnet/roslyn/issues/39166
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: No, it cannot be.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs [hunk 2]
```diff
@@ -186,7 +188,9 @@ private CustomAttributesBag<CSharpAttributeData> GetAttributesBag()
                 Debug.Assert(wasCompletedThisThread);
             }
 
+#nullable disable // Can '_lazyCustomAttributesBag' be null? https://github.com/dotnet/roslyn/issues/39166
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: No. `LoadAndValidateAttributes` initializes the parameter always, except when `earlyDecodingOnly` is true. It is not in the call above.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs [hunk 3]
```diff
@@ -292,23 +296,27 @@ internal sealed override void DecodeWellKnownAttribute(ref DecodeWellKnownAttrib
             else if (attribute.IsTargetAttribute(this, AttributeDescription.NullableAttribute))
             {
                 // NullableAttribute should not be set explicitly.
+#nullable disable // Can 'arguments.AttributeSyntaxOpt' be null? https://github.com/dotnet/roslyn/issues/39166
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: No. This is decoding a source-based attribute: if it is of the attribute type, it's not null.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: Same for below.

---
In reply to: [364968058](https://github.com/dotnet/roslyn/pull/39194#discussion_r364968058) [](ancestors = 364968058)

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationEventSymbol.cs [hunk 4]
```diff
@@ -50,14 +53,19 @@ protected override CodeGenerationSymbol Clone()
         public override void Accept(SymbolVisitor visitor)
             => visitor.VisitEvent(this);
 
+        [return: MaybeNull]
         public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
-            => visitor.VisitEvent(this);
+        {
+#pragma warning disable CS8717 // A member returning a [MaybeNull] value introduces a null value when 'TResult' is a non-nullable reference type.
```
Comment (333fred) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationEventSymbol.cs]: Has this warning not been removed yet? I didn't think it shipped in 16.4.

**Comment by 333fred** [comment]
Done review pass (commit 3)a

**Comment by 333fred** [comment]
@sharwell looks like there's a merge conflict, and I don't see responses to my other comments.

**Comment by sharwell** [comment]
@333fred I have not completed my updates to this one yet

**Comment by 333fred** [comment]
Alright, ping me when you're ready.

File: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: I'm planning to leave this one for a follow-up

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: I'm planning to leave this one for a follow-up

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: I'm planning to leave this one for a follow-up

**Comment by sharwell** [comment]
@333fred this is now ready for a new review.

File: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: Why? The question has been answered? Same for all the other comments that I've answered in this PR.

---
In reply to: [368006450](https://github.com/dotnet/roslyn/pull/39194#discussion_r368006450) [](ancestors = 368006450)

**Comment by 333fred** [comment]
Done review pass (commit 6)

File: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: That was the original intent of all the `#nullable disable` lines in all the PRs I submitted. The comment was just added for clarity for whoever looks at these later during a cleanup pass.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: Sure, but I've taken the time and answered the questions, and I don't really want to leave these in a half-baked state.

Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: I'll get the ones in this pull request updated, but keep in mind for the ones that are still pending review?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: Alright.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: Updated

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs]: Updated both

File: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: Updated

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler  for a second review

File: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs [hunk 5]
```diff
@@ -258,14 +261,15 @@ internal EventSymbol AsMember(NamedTypeSymbol newOwner)
         {
             Debug.Assert(this.IsDefinition);
             Debug.Assert(ReferenceEquals(newOwner.OriginalDefinition, this.ContainingSymbol.OriginalDefinition));
-            return newOwner.IsDefinition ? this : new SubstitutedEventSymbol(newOwner as SubstitutedNamedTypeSymbol, this);
+            Debug.Assert(newOwner.IsDefinition || newOwner is SubstitutedNamedTypeSymbol);
+            return newOwner.IsDefinition ? this : new SubstitutedEventSymbol((newOwner as SubstitutedNamedTypeSymbol)!, this);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: >(newOwner as SubstitutedNamedTypeSymbol)! [](start = 77, length = 41)

Should be a cast then?

**Comment by jcouv** [comment]
>                 SyntaxReference reference = symbol.DeclaringSyntaxReferences.FirstOrDefault();

`?` since being null-tested on next line

---
Refers to: src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:320 in 902cb5a. [](commit_id = 902cb5a87c2596106fba589e5a86c7d762b69938, deletion_comment = False)

**Comment by jcouv** [comment]
>                     Symbol containingSymbol = symbol.ContainingSymbol;

`?` since null-tested below

---
Refers to: src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:324 in 902cb5a. [](commit_id = 902cb5a87c2596106fba589e5a86c7d762b69938, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEEventSymbol.cs [hunk 6]
```diff
@@ -20,16 +22,18 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
     /// </summary>
     internal sealed class PEEventSymbol : EventSymbol
     {
+#nullable disable // The compiler cannot figure out that this is assigned in the constructor. https://github.com/dotnet/roslyn/issues/39166
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEEventSymbol.cs]: Tagging @jaredpar for nullable guideline: Should we use `#nullable disable` or `= null!;` for fields that are initialized later.

**Comment by jcouv** [comment]
>             BadImageFormatException mrEx;

`?` since null-tested below

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEEventSymbol.cs:442 in 902cb5a. [](commit_id = 902cb5a87c2596106fba589e5a86c7d762b69938, deletion_comment = False)

**Comment by jcouv** [comment]
>             string aliasQualifierOpt;

Should this be `?` ?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Source/SourceCustomEventSymbol.cs:35 in 902cb5a. [](commit_id = 902cb5a87c2596106fba589e5a86c7d762b69938, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs (hunk 5 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: I couldn't confirm statically that this conversion is always non-null, and didn't want to be the cause of a new exception.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/EventSymbol.cs]: ➡️ okay

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEEventSymbol.cs (hunk 6 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEEventSymbol.cs]: Reported #41273 and applied a workaround in code

**Comment by sharwell** [comment]
@jcouv Updated all cases.

**Comment by jcouv** [comment]
@sharwell PR could be merged (Fred and I signed off)

**Comment by jcouv** [comment]
@sharwell PR could be merged

**Pull request merged**